devbox/src/cmdproxy/ProxyServer.scala

```diff
@@ -0,0 +1,85 @@
+package cmdproxy
+
+import java.io.BufferedReader
+import java.io.InputStreamReader
+import java.io.OutputStreamWriter
+import java.io.PrintWriter
+import java.net.InetAddress
+import java.net.ServerSocket
+import java.net.Socket
+
+import scala.util.Try
+
+import devbox.logger.SyncLogger
+import os.RelPath
+
+case class Request(verb: String, workingDir: String, args: Seq[String])
+
+/**
+ * Simple server that runs commands on the local system and responds with the output of that command.
+ *
+ * The server is single-threaded and synchronous: each request is handled before the next one is accepted.
+ *
+ * Protocol is based on UTF-8 encoded text:
+ *
+ * Request is one line, terminated with \n: RUN dir cmdline
+ * Response:
+ *   EXITCODE: N
+ *   output
+ */
+class ProxyServer(dirMapping: Seq[(os.Path, os.RelPath)], port: Int = ProxyServer.DEFAULT_PORT)(implicit logger: SyncLogger) {
+  val socket = new ServerSocket(port, 2, InetAddress.getLoopbackAddress)
+
+  /** Map remote directories to local ones. */
+  val localDir: Map[os.RelPath, os.Path] = dirMapping.map(_.swap).toMap
+
+  def start(): Unit = {
+    logger.info(s"Starting command proxy server, listening at ${socket.getInetAddress}:${socket.getLocalPort}")
+    while (true) {
+      val newConnection = socket.accept()
+      handleConnection(newConnection)
+    }
+  }
+
+  def handleConnection(conn: Socket): Unit = try {
+    logger.info(s"Accepting connection from ${conn.getInetAddress}")
+    val in = new BufferedReader(new InputStreamReader(conn.getInputStream, ProxyServer.CHARSET_NAME))
+    val out = new PrintWriter(new OutputStreamWriter(conn.getOutputStream, ProxyServer.CHARSET_NAME))
+
+    parseRequest(in.readLine()) match {
+      case Some(Request("RUN", dir, args)) =>
+        val proc = os.proc(args).call(
+          localDir.getOrElse(RelPath(dir), os.home / RelPath(dir)),
+          mergeErrIntoOut = true,
+          check = false,
+          timeout = 10000)
+
+        out.println(s"EXITCODE: ${proc.exitCode}")
+        out.println(proc.out.string(ProxyServer.CHARSET_NAME))
+
+        // we don't keep alive the connection at this time
+        out.close()
+
+      case _ => ()
+    }
+  } catch {
+    case e: Exception =>
+      logger.error(s"Error handling connection from ${conn.getInetAddress}", e.getMessage)
+      Try(conn.close())
+  }
```

Comment:
Should there be a `finally` block here that closes the connection no matter what happens? Could also move the `out.close()` above into this `finally` block